use findbin::realbin so the scripts can be symlinked to the user path

warmfusion opened the pull request:
I wanted to install these scripts to /usr/local/src/myq_gadgets and then symlink to the scripts themselves into /usr/local/bin/ but the symlink resulted in a failure to resolve the mysql_script_utils library as the lib path was incorrect.

I've changed to realbin which resolves the 'real' location of the script, enabling symlinked scripts to work as expected.

warmfusion committed:
use findbin::realbin so that the scripts can be installed in one place, symlinked into /usr/local/bin and executed

jayjanssen commented:
I'm not familiar with Find::RealBin -- how far back into Perl's history does that go?  One advantage I want to maintain here is that it runs virtually with any version of Perl.

warmfusion commented:
Thats understandable, I'm found that the realbin variable was added in 2011 in (1.52 of the Findbin libs )[http://cpansearch.perl.org/src/LEMBARK/FindBin-libs-2.12/Changes]

Specifically look for:

> 1.52 Fri Sep  9 17:10:29 EDT 2011

I could add a defined check and use realbin if it exists, and failback to bin otherwise?

jayjanssen commented:
I would be ok with that.

On Tue, Jan 27, 2015 at 3:34 PM, Toby Jackson <notifications@github.com>
wrote:

> Thats understandable, I'm found that the realbin variable was added in
> 2011 in (1.52 of the Findbin libs )[
> http://cpansearch.perl.org/src/LEMBARK/FindBin-libs-2.12/Changes]
>
> Specifically look for:
>
> 1.52 Fri Sep 9 17:10:29 EDT 2011
>
> I could add a defined check and use realbin if it exists, and failback to
> bin otherwise?
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/jayjanssen/myq_gadgets/pull/9#issuecomment-71722538>.
>



-- 
Jay Janssen
Managing Consultant, Percona
https://about.me/jay.janssen

warmfusion committed:
Defensive FindBin::Bin lookup for Perl < 5.10

Try and use FindBin::realbin if it returns a value, but otherwise default to findbin::bin
as that should always have a value to use.

warmfusion commented:
Ok, I've had a go, and 'tested' by doing some rudimentary checks through my Perl 5.18 installation. It works quite happily and behaves as i'd expect when the scripts were symlinked into /tmp

There are four test cases testing the combination of symlinked/not and with a value returned by RealBin and not, the output is shown below

## Valid RealBin and Real Script
````
Tobys-MacBook-Pro:myq_gadgets toby$ ./myq_status
'mysql' binary not found in your $PATH
````

## Valid RealBin and Symlinked
````
...
lrwxr-xr-x   1 toby  wheel   44 27 Jan 21:48 myq_status -> /Users/toby/workspace/myq_gadgets/myq_status
...
Tobys-MacBook-Pro:tmp toby$ ./myq_status
'mysql' binary not found in your $PATH
````

## Invalid RealBin and Symlinked
If I change RealBin to RealBin2, a variable that is not defined in FindBin, the script outputs this instead:

````
Tobys-MacBook-Pro:tmp toby$ ./myq_status
Can't locate MySQL_Script_Utils.pm in @INC (you may need to install the MySQL_Script_Utils module) (@INC contains: /private/tmp /Library/Perl/5.18/darwin-thread-multi-2level /Library/Perl/5.18 /Network/Library/Perl/5.18/darwin-thread-multi-2level /Network/Library/Perl/5.18 /Library/Perl/Updates/5.18.2 /System/Library/Perl/5.18/darwin-thread-multi-2level /System/Library/Perl/5.18 /System/Library/Perl/Extras/5.18/darwin-thread-multi-2level /System/Library/Perl/Extras/5.18 .) at ./myq_status line 33.
````

Note how it indicates /private/tmp is included on the path - this shows that FindBin::bin is still being used if RealBin wouldn't return a value, and does so quietly (No warnings shown in the console output)


## Invalid RealBin and real script
Finally, when running from its real location, still with a "RealBin2" configured, the script works as normal

````
Tobys-MacBook-Pro:myq_gadgets toby$ ./myq_status
'mysql' binary not found in your $PATH
````

jayjanssen commented:
Nice work, thanks.  If I have any issues, I'll just rip it out later and come back to you ;)

jayjanssen commented:
Somewhat ironically, I'm focusing more effort into https://github.com/jayjanssen/myq-tools/ these days, but that isn't feature complete yet compared with myq_gadgets
